File: src/app/model/SpeedSensorData.ts
export class SpeedSensorData {
    constructor(
        public timestamp: Date,
        public cumulativeWheelRevolutions?: number, // Cumulative wheel revolutions Uint32
        public cumulativeCrankRevolutions?: number, // Cumulative crank revolutions Uint16
        public lastWheelEventTime?: number, // Last wheel event time Uint16 in 1/1024 second
        public lastCrankEventTime?: number // Last crank event time Uint16 in 1/1024 second
    ) { }

    public diffWheelRevolutions(previous: SpeedSensorData): number | null {
        return this.diff(this.cumulativeWheelRevolutions, previous.cumulativeWheelRevolutions, 0xFFFFFFFF);
    }

    public diffCrankRevolutions(previous: SpeedSensorData): number | null {
        return this.diff(this.cumulativeCrankRevolutions, previous.cumulativeCrankRevolutions, 0xFFFF);
    }

    public diffWheelTimeInSeconds(previous: SpeedSensorData): number | null {
        const diff = this.diff(this.lastWheelEventTime, previous.lastWheelEventTime, 0xFFFF)
        if (diff === null) {
            return null;
        }
        return diff / 1024;
    }
    public diffCrankTimeInSeconds(previous: SpeedSensorData): number | null {
        const diff = this.diff(this.lastCrankEventTime, previous.lastCrankEventTime, 0xFFFF);
        if (diff === null) {
            return null;
        }
        return diff / 1024;
    }

    private diff(biggest: number | undefined, smallest: number | undefined, maxValue: number): number | null {
        if (biggest === undefined || smallest === undefined) {
            return null;
        }
        if (biggest < smallest) {
            return (biggest + maxValue) - smallest;
        }
        return biggest - smallest;
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/SpeedSensorData.java
package com.openbikecomputer.model;

/**
 * Parsed Bluetooth Cycling Speed and Cadence (CSC) measurement.
 * Ported from {@code SpeedSensorData.ts}.
 */
public class SpeedSensorData {
    /** Epoch milliseconds. */
    public long timestamp;
    /** Cumulative wheel revolutions, Uint32. */
    public Long cumulativeWheelRevolutions;
    /** Cumulative crank revolutions, Uint16. */
    public Long cumulativeCrankRevolutions;
    /** Last wheel event time, Uint16 in 1/1024 second. */
    public Long lastWheelEventTime;
    /** Last crank event time, Uint16 in 1/1024 second. */
    public Long lastCrankEventTime;

    public SpeedSensorData(long timestamp) {
        this.timestamp = timestamp;
    }

    public Double diffWheelRevolutions(SpeedSensorData previous) {
        return diff(cumulativeWheelRevolutions, previous.cumulativeWheelRevolutions, 0xFFFFFFFFL);
    }

    public Double diffCrankRevolutions(SpeedSensorData previous) {
        return diff(cumulativeCrankRevolutions, previous.cumulativeCrankRevolutions, 0xFFFFL);
    }

    public Double diffWheelTimeInSeconds(SpeedSensorData previous) {
        Double d = diff(lastWheelEventTime, previous.lastWheelEventTime, 0xFFFFL);
        if (d == null) {
            return null;
        }
        return d / 1024.0;
    }

    public Double diffCrankTimeInSeconds(SpeedSensorData previous) {
        Double d = diff(lastCrankEventTime, previous.lastCrankEventTime, 0xFFFFL);
        if (d == null) {
            return null;
        }
        return d / 1024.0;
    }

    private static Double diff(Long biggest, Long smallest, long maxValue) {
        if (biggest == null || smallest == null) {
            return null;
        }
        if (biggest < smallest) {
            return (double) ((biggest + maxValue) - smallest);
        }
        return (double) (biggest - smallest);
    }
}
